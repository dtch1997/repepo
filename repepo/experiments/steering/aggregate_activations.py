from repepo.experiments.steering.utils.helpers import (
    SteeringConfig,
    load_activations,
    save_concept_vectors,
)
from repepo.experiments.steering.utils.configs import list_configs

from steering_vectors.aggregators import (
    Aggregator,
    mean_aggregator,
    logistic_aggregator,
    pca_aggregator,
)
from steering_vectors.train_steering_vector import aggregate_activations

aggregators = {
    "mean": mean_aggregator,
    "logistic": logistic_aggregator,
    "pca": pca_aggregator,
}


def run_aggregate_activations(config: SteeringConfig):
    pos_acts = load_activations(config, "positive")
    neg_acts = load_activations(config, "negative")
    aggregator: Aggregator = aggregators[config.aggregator]()
    concept_vectors = aggregate_activations(pos_acts, neg_acts, aggregator)
    save_concept_vectors(config, concept_vectors)
    return concept_vectors


if __name__ == "__main__":
    import simple_parsing

    parser = simple_parsing.ArgumentParser()
    parser.add_arguments(SteeringConfig, dest="config")
    parser.add_argument("--configs", type=str, default="")
    args = parser.parse_args()
    config = args.config

    if args.configs:
        configs = list_configs(
            args.configs, config.train_split_name, config.test_split_name
        )
        for config in configs:
            print(f"Running on dataset: {config.train_dataset_name}")
            run_aggregate_activations(config)

    else:
        run_aggregate_activations(config)
