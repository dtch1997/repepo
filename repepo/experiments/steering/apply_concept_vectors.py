import torch
import simple_parsing


from repepo.experiments.steering.utils.helpers import (
    get_model_name,
    get_model_and_tokenizer,
    get_formatter,
    SteeringConfig,
    SteeringResult,
    load_concept_vectors,
    save_results,
    make_dataset,
    EmptyTorchCUDACache,
)
from repepo.experiments.steering.utils.configs import list_configs

from repepo.core.types import Dataset
from repepo.core.pipeline import Pipeline
from repepo.core.evaluate import (
    print_first_example,
    select_repe_layer_at_eval,
    set_repe_direction_multiplier_at_eval,
    evaluate,
    MultipleChoiceAccuracyEvaluator,
    LogitDifferenceEvaluator,
    NormalizedPositiveProbabilityEvaluator,
)
from repepo.core.hook import SteeringHook
from steering_vectors import SteeringVector


def evaluate_steering_with_concept_vectors(
    pipeline: Pipeline,
    concept_vectors: dict[int, torch.Tensor],
    dataset: Dataset,
    layers: list[int],
    multipliers: list[float],
    verbose: bool = False,
) -> list[SteeringResult]:
    caa_results = []

    steering_vector = SteeringVector(layer_activations=concept_vectors)

    # Create steering hook and add it to pipeline
    steering_hook = SteeringHook(
        steering_vector=steering_vector,
        direction_multiplier=0,
        patch_generation_tokens_only=False,
        skip_first_n_generation_tokens=0,
        layer_config=None,
    )
    pipeline.hooks.append(steering_hook)

    for layer_id in layers:
        for multiplier in multipliers:
            pass
            # Run evaluate to get metrics
            result = evaluate(
                pipeline,
                dataset,
                eval_hooks=[
                    # TODO: different datasets need different of evaluating prompt template
                    # update_completion_template_at_eval(
                    #     "{prompt} My answer is {response}"
                    # ),
                    print_first_example(),
                    set_repe_direction_multiplier_at_eval(multiplier),
                    select_repe_layer_at_eval(layer_id),
                ],
                evaluators=[
                    MultipleChoiceAccuracyEvaluator(),
                    LogitDifferenceEvaluator(),
                    NormalizedPositiveProbabilityEvaluator(),
                ],
                verbose=True,
            )

            caa_result = SteeringResult(
                layer_id=layer_id,
                multiplier=multiplier,
                mcq_acc=result.metrics["mcq_acc"],
                logit_diff=result.metrics["logit_diff"],
                pos_prob=result.metrics["pos_prob"],
            )
            caa_results.append(caa_result)
            if verbose:
                print(
                    f"Layer {layer_id}, multiplier {multiplier:.2f}: "
                    f"MCQ Accuracy {caa_result.mcq_acc:.2f} "
                    f"Positive Prob {caa_result.pos_prob:.2f} "
                    f"Logit Diff {caa_result.logit_diff:.2f} "
                )

    return caa_results


def run_load_extract_and_save(
    config: SteeringConfig,
):
    test_dataset = make_dataset(config.test_dataset_name, config.test_split_name)
    model_name = get_model_name(config.use_base_model, config.model_size)
    model, tokenizer = get_model_and_tokenizer(model_name)
    formatter = get_formatter(config.formatter)

    if len(config.layers) == 0:
        config.layers = list(range(model.config.num_hidden_layers))
    pipeline = Pipeline(model, tokenizer, formatter)
    concept_vectors = load_concept_vectors(config)

    results = evaluate_steering_with_concept_vectors(
        pipeline=pipeline,
        concept_vectors=concept_vectors,
        dataset=test_dataset,
        layers=config.layers,
        multipliers=config.multipliers,
        verbose=config.verbose,
    )

    save_results(config, results)


if __name__ == "__main__":
    parser = simple_parsing.ArgumentParser()
    parser.add_arguments(SteeringConfig, dest="config")
    parser.add_argument("--configs", type=str, default="")

    args = parser.parse_args()
    config = args.config

    if args.configs:
        all_configs = list_configs(
            args.configs, config.train_split_name, config.test_split_name
        )
        for config in all_configs:
            # Train and test on the same dataset
            print("Running on dataset: ", config.test_dataset_name)
            with EmptyTorchCUDACache():
                run_load_extract_and_save(config)
    else:
        with EmptyTorchCUDACache():
            run_load_extract_and_save(config)
