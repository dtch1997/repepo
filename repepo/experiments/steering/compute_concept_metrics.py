import torch
from torch import Tensor
from torchmetrics.functional import pairwise_cosine_similarity
from typing import Callable

from repepo.experiments.steering.utils.helpers import (
    list_subset_of_datasets,
    get_configs_for_datasets,
    ConceptVectorsConfig,
    load_activations,
    compute_difference_vectors,
    save_concept_vectors,
    save_metrics,
)

Aggregator = Callable[[list[Tensor]], Tensor]
Metric = Callable[[list[Tensor]], float]


def _mean_agg(diff_vecs: list[Tensor]) -> Tensor:
    """Compute the mean over the list of difference vectors."""
    return torch.mean(torch.stack(diff_vecs), dim=0)


def _mean_norm(diff_vecs: list[Tensor]) -> float:
    """Compute the mean norm of the difference vectors."""
    return torch.mean(torch.stack([torch.norm(v) for v in diff_vecs])).item()


def _mean_var_scaled_by_norm_of_mean_of_diff_agg(diff_vecs: list[Tensor]) -> float:
    """Compute the mean variance of the difference vectors, scaled by the norm of the mean difference vector"""
    mean = _mean_agg(diff_vecs)
    vecs_to_mean = [diff_vec - mean for diff_vec in diff_vecs]
    mean_norm = torch.norm(mean)
    norm_vecs_to_mean = [torch.norm(vec) for vec in vecs_to_mean]
    return (torch.mean(torch.stack(norm_vecs_to_mean)) / mean_norm).item()


def _mean_var_scaled_by_mean_of_norm_of_diff_agg(diff_vecs: list[Tensor]) -> float:
    """Compute the mean variance of the difference vectors, scaled by the mean of the norms of the difference vectors."""
    mean = _mean_agg(diff_vecs)
    vecs_to_mean = [diff_vec - mean for diff_vec in diff_vecs]
    norm_vecs_to_mean = [torch.norm(vec) for vec in vecs_to_mean]
    return (torch.mean(torch.stack(norm_vecs_to_mean)) / _mean_norm(diff_vecs)).item()


def _mean_var_unscaled_agg(diff_vecs: list[Tensor]) -> float:
    """Compute the mean unscaled variance of the difference vectors."""
    mean = _mean_agg(diff_vecs)
    vecs_to_mean = [diff_vec - mean for diff_vec in diff_vecs]
    norm_vecs_to_mean = [torch.norm(vec) for vec in vecs_to_mean]
    return torch.mean(torch.stack(norm_vecs_to_mean)).item()


def _mean_pairwise_cosine_agg(diff_vecs: list[Tensor]) -> float:
    """Compute the mean cosine similarity between each pair of difference vectors."""
    diff_vecs_tensor = torch.stack(diff_vecs)
    pairwise_cosine_sims = pairwise_cosine_similarity(
        diff_vecs_tensor, diff_vecs_tensor
    )
    return torch.mean(pairwise_cosine_sims).item()


def _mean_pairwise_dot_product_agg(diff_vecs: list[Tensor]) -> float:
    """Compute the mean dot product between each pair of difference vectors."""
    diff_vecs_tensor = torch.stack(diff_vecs)
    pairwise_dot_products = torch.einsum(
        "id,jd->ij", diff_vecs_tensor, diff_vecs_tensor
    )
    return torch.mean(pairwise_dot_products).item()


metrics: dict[str, Metric] = {
    "mean_norm": _mean_norm,
    "var_mean_norm_diff": _mean_var_scaled_by_mean_of_norm_of_diff_agg,
    "var_norm_mean_diff": _mean_var_scaled_by_norm_of_mean_of_diff_agg,
    "unscaled_var": _mean_var_unscaled_agg,
    "pairwise_cosine": _mean_pairwise_cosine_agg,
    "pairwise_dot_product": _mean_pairwise_dot_product_agg,
}


def aggregate(
    activations: dict[int, list[Tensor]], aggregator: Aggregator
) -> dict[int, Tensor]:
    aggs: dict[int, Tensor] = {}
    for layer_num in activations.keys():
        diff_vecs = activations[layer_num]
        aggs[layer_num] = aggregator(diff_vecs)
    return aggs


def run_load_extract_and_save(config: ConceptVectorsConfig, metric_names: list[str]):
    print("Running on dataset: ", config.train_dataset_name)
    positive_activations = load_activations(config, "positive")
    negative_activations = load_activations(config, "negative")
    difference_vectors = compute_difference_vectors(
        positive_activations, negative_activations
    )

    # Squeeze
    for layer_num in difference_vectors.keys():
        difference_vectors[layer_num] = [
            diff_vec.squeeze(0) for diff_vec in difference_vectors[layer_num]
        ]

    concept_vectors = aggregate(difference_vectors, aggregator=_mean_agg)
    save_concept_vectors(config, concept_vectors)
    for metric_name in metric_names:
        metric = metrics[metric_name]
        metric_val = {
            layer_num: metric(diff_vecs)
            for layer_num, diff_vecs in difference_vectors.items()
        }
        save_metrics(config, metric_name, metric_val)  # type: ignore


if __name__ == "__main__":
    import simple_parsing

    parser = simple_parsing.ArgumentParser()
    parser.add_arguments(ConceptVectorsConfig, dest="config")
    parser.add_argument("--datasets", type=str, default="")
    # parser.add_argument("--aggregator", type=str, default="mean")
    args = parser.parse_args()
    config = args.config

    metric_names = list(metrics.keys())

    if args.datasets:
        all_datasets = list_subset_of_datasets(args.datasets)
        configs = get_configs_for_datasets(all_datasets, config.train_split_name)
        for config in configs:
            run_load_extract_and_save(config, metric_names=metric_names)
    else:
        run_load_extract_and_save(config, metric_names=metric_names)
