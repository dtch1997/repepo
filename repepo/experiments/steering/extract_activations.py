import torch
from repepo.core.types import Dataset
from repepo.core.format import LlamaChatFormatter
from repepo.experiments.steering.utils.helpers import (
    get_model_name,
    get_model_and_tokenizer,
    ConceptVectorsConfig,
    save_activations,
    list_subset_of_datasets,
    get_configs_for_datasets,
    make_dataset,
    EmptyTorchCUDACache,
)

from steering_vectors.train_steering_vector import extract_activations


def _validate_train_dataset(dataset: Dataset):
    for example in dataset:
        assert example.steering_token_index == -2


@torch.no_grad()
def run_extract_activations(
    config: ConceptVectorsConfig,
):
    model_name = get_model_name(config.use_base_model, config.model_size)
    model, tokenizer = get_model_and_tokenizer(model_name)
    formatter = LlamaChatFormatter()

    train_dataset = make_dataset(config.train_dataset_name, config.train_split_name)
    _validate_train_dataset(train_dataset)

    repe_training_data = [
        (
            formatter.format_as_str(formatter.format_conversation(example.positive)),
            formatter.format_as_str(formatter.format_conversation(example.negative)),
        )
        for example in train_dataset
    ]

    _pos, _neg = repe_training_data[0]
    print(_pos)
    print(_neg)

    # Extract activations
    pos_acts, neg_acts = extract_activations(
        model,
        tokenizer,
        repe_training_data,
        read_token_index=-2,
        show_progress=True,
        move_to_cpu=True,
    )

    return pos_acts, neg_acts


def run_extract_and_save(config: ConceptVectorsConfig):
    with EmptyTorchCUDACache():
        pos, neg = run_extract_activations(config)
        save_activations(config, pos, "positive")
        save_activations(config, neg, "negative")


if __name__ == "__main__":
    import simple_parsing

    parser = simple_parsing.ArgumentParser()
    parser.add_arguments(ConceptVectorsConfig, dest="config")
    parser.add_argument("--datasets", type=str, default="")
    args = parser.parse_args()
    config = args.config

    if args.datasets:
        all_datasets = list_subset_of_datasets(args.datasets)
        configs = get_configs_for_datasets(all_datasets, config.train_split_name)
        for config in configs:
            print(f"Running on dataset: {config.train_dataset_name}")
            run_extract_and_save(config)

    else:
        run_extract_and_save(config)
