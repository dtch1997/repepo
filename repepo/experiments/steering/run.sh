#!/bin/bash
configs_arg="${1:-dev}"
train_split_arg="${2:-train-dev}"
test_split_arg="${3:-val-dev}"

echo "Running steering experiments with configs: $configs_arg, train_split: $train_split_arg, test_split: $test_split_arg"

python repepo/experiments/steering/extract_activations.py \
    --configs "$configs_arg" \
    --train_split_name "$train_split_arg" \
    --test_split_name "$test_split_arg"    

python repepo/experiments/steering/compute_concept_metrics.py \
    --configs "$configs_arg" \
    --train_split_name "$train_split_arg" \
    --test_split_name "$test_split_arg"

python repepo/experiments/steering/aggregate_activations.py \
    --configs "$configs_arg" \
    --train_split_name "$train_split_arg" \
    --test_split_name "$test_split_arg"

python repepo/experiments/steering/apply_concept_vectors.py \
    --configs "$configs_arg" \
    --train_split_name "$train_split_arg" \
    --test_split_name "$test_split_arg"
