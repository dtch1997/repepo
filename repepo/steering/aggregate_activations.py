from torch import Tensor
from steering_vectors.aggregators import (
    Aggregator,
    mean_aggregator,
    logistic_aggregator,
    pca_aggregator,
)
from steering_vectors.train_steering_vector import (
    aggregate_activations as _aggregate_activations,
)

aggregators = {
    "mean": mean_aggregator,
    "logistic": logistic_aggregator,
    "pca": pca_aggregator,
}


def get_aggregator(name: str) -> Aggregator:
    """A wrapper around steering_vectors.aggregators.get_aggregator"""
    return aggregators[name]()


def aggregate_activations(
    positive_activations: dict[int, list[Tensor]],
    negative_activations: dict[int, list[Tensor]],
    aggregator: Aggregator,
    verbose: bool = False,
):
    """A wrapper around steering_vectors.train_steering_vector.aggregate_activations"""
    return _aggregate_activations(
        positive_activations, negative_activations, aggregator
    )
