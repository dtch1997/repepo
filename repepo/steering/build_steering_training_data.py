import logging
from repepo.core.types import Dataset
from repepo.core.pipeline import Pipeline
from steering_vectors import SteeringVectorTrainingSample


def _validate_train_dataset(dataset: Dataset):
    steering_token_index = dataset[0].steering_token_index
    for example in dataset:
        assert example.steering_token_index == steering_token_index


def build_steering_vector_training_data(
    pipeline: Pipeline,
    dataset: Dataset,
    logger: logging.Logger | None = None,
) -> list[SteeringVectorTrainingSample]:
    # Validate that all examples have the same steering token index
    _validate_train_dataset(dataset)
    # After validation, we can assume that all examples have the same steering token index
    read_token_index = dataset[0].steering_token_index

    # NOTE(dtch1997): Using SteeringVectorTrainingSample here
    # to encode information about token index
    steering_vector_training_data = [
        SteeringVectorTrainingSample(
            positive_str=pipeline.build_full_prompt(example.positive),
            negative_str=pipeline.build_full_prompt(example.negative),
            read_positive_token_index=read_token_index,
            read_negative_token_index=read_token_index,
        )
        for example in dataset
    ]

    if logger is not None:
        # Log first example
        datum = steering_vector_training_data[0]
        logger.info(f"Positive example: \n {datum.positive_str}")
        logger.info(f"Negative example: \n {datum.negative_str}")

    return steering_vector_training_data
