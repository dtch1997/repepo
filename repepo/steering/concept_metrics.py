import torch
from torch import Tensor
from torchmetrics.functional import pairwise_cosine_similarity
from typing import Protocol


def _mean_vector(diff_vecs: list[Tensor]) -> Tensor:
    """Compute the mean over the list of difference vectors."""
    return torch.mean(torch.cat(diff_vecs), dim=0)


def _norm_mean_vector(diff_vecs: list[Tensor]) -> float:
    """Compute the norm of the mean difference vector."""
    mean = _mean_vector(diff_vecs)
    return torch.norm(mean).item()


def _mean_norm(diff_vecs: list[Tensor]) -> float:
    """Compute the mean norm of the difference vectors."""
    return torch.mean(torch.stack([torch.norm(v) for v in diff_vecs])).item()


def _var_norm(diff_vecs: list[Tensor]) -> float:
    """Compute the variance of the norms of the difference vectors."""
    return torch.var(torch.stack([torch.norm(v) for v in diff_vecs])).item()


def _var_norm_scaled_by_mean_norm(diff_vecs: list[Tensor]) -> float:
    return _var_norm(diff_vecs) / _mean_norm(diff_vecs)


def _mean_euclidean_distance(diff_vecs: list[Tensor]) -> float:
    """Compute the average distance of the difference vectors to the mean vector."""
    mean_vector = _mean_vector(diff_vecs)
    distances = [diff_vec - mean_vector for diff_vec in diff_vecs]
    norm_vecs_to_mean = [torch.norm(vec) for vec in distances]
    return torch.mean(torch.stack(norm_vecs_to_mean)).item()


def _mean_euclidean_distance_scaled_by_norm_mean_vector(
    diff_vecs: list[Tensor],
) -> float:
    return _mean_euclidean_distance(diff_vecs) / _norm_mean_vector(diff_vecs)


def _mean_euclidean_distance_scaled_by_mean_norm(diff_vecs: list[Tensor]) -> float:
    return _mean_euclidean_distance(diff_vecs) / _mean_norm(diff_vecs)


def _mean_cosine_similarity(diff_vecs: list[Tensor]) -> float:
    """Compute the mean cosine similarity between each pair of difference vectors."""
    diff_vecs_tensor = torch.cat(diff_vecs)
    pairwise_cosine_sims = pairwise_cosine_similarity(
        diff_vecs_tensor, diff_vecs_tensor
    )
    return torch.mean(pairwise_cosine_sims).item()


def _mean_dot_product(diff_vecs: list[Tensor]) -> float:
    """Compute the mean dot product between each pair of difference vectors."""
    diff_vecs_tensor = torch.cat(diff_vecs)
    pairwise_dot_products = torch.einsum(
        "id,jd->ij", diff_vecs_tensor, diff_vecs_tensor
    )
    return torch.mean(pairwise_dot_products).item()


class Metric(Protocol):
    def __call__(self, diff_vecs: list[Tensor]) -> float:
        ...


def compute_difference_vectors(
    pos_acts: list[Tensor], neg_acts: list[Tensor]
) -> list[Tensor]:
    return [pos - neg for pos, neg in zip(pos_acts, neg_acts)]


class VarianceOfNormSimilarityMetric:
    name: str = "variance_of_norm_sim"

    def __call__(self, diff_vecs) -> float:
        return -1.0 * _var_norm_scaled_by_mean_norm(diff_vecs)


class EuclideanSimilarityMetric:
    name: str = "euclidean_sim"

    def __call__(self, diff_vecs: list[Tensor]) -> float:
        return -1.0 * _mean_euclidean_distance_scaled_by_mean_norm(diff_vecs)


class CosineSimilarityMetric:
    name: str = "cosine_sim"

    def __call__(self, diff_vecs: list[Tensor]) -> float:
        return _mean_cosine_similarity(diff_vecs)
