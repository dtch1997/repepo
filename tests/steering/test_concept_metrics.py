import torch
from repepo.steering.concept_metrics import (
    VarianceOfNormSimilarityMetric,
    EuclideanSimilarityMetric,
    CosineSimilarityMetric,
    compute_difference_vectors,
)

from syrupy import SnapshotAssertion


def test_compute_difference_vectors():
    pos_acts = [
        torch.tensor([[1, 2, 3]], dtype=torch.float32),
        torch.tensor([[4, 5, 6]], dtype=torch.float32),
    ]
    neg_acts = [
        torch.tensor([[7, 8, 9]], dtype=torch.float32),
        torch.tensor([[10, 11, 12]], dtype=torch.float32),
    ]
    diff_vecs = compute_difference_vectors(pos_acts, neg_acts)
    target_diff_vecs = [
        torch.tensor([[-6, -6, -6]], dtype=torch.float32),
        torch.tensor([[-6, -6, -6]], dtype=torch.float32),
    ]

    for diff_vec, target_diff_vec in zip(diff_vecs, target_diff_vecs):
        assert torch.allclose(diff_vec, target_diff_vec, atol=1e-5)


def test_cosine_similarity_metric(snapshot: SnapshotAssertion):
    metric = CosineSimilarityMetric()
    diff_vecs = [
        torch.tensor([[1, 2, 3]], dtype=torch.float32),
        torch.tensor([[4, 5, 6]], dtype=torch.float32),
    ]
    assert metric(diff_vecs) == snapshot


def test_euclidean_similarity_metric(snapshot: SnapshotAssertion):
    metric = EuclideanSimilarityMetric()
    diff_vecs = [
        torch.tensor([[1, 2, 3]], dtype=torch.float32),
        torch.tensor([[4, 5, 6]], dtype=torch.float32),
    ]
    assert metric(diff_vecs) == snapshot


def test_variance_of_norm_similarity_metric(snapshot: SnapshotAssertion):
    metric = VarianceOfNormSimilarityMetric()
    diff_vecs = [
        torch.tensor([[1, 2, 3]], dtype=torch.float32),
        torch.tensor([[4, 5, 6]], dtype=torch.float32),
    ]
    assert metric(diff_vecs) == snapshot
