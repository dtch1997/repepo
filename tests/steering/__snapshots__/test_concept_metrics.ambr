# serializer version: 1
# name: test_cosine_similarity_metric
  0.9873158931732178
# ---
# name: test_euclidean_similarity_metric
  -0.41514015367596296
# ---
# name: test_variance_of_norm_similarity_metric
  -2.0240427367508826
# ---
